File: ml/dataset/build_dataset.py
import random

from .build_user_tensor import build_user_tensor

def build_dataset(
    sequence,
    all_songs,
    sequence_length=10,
    negative_per_positive=1,
    embedding_length=None
):
    samples = []

    if not sequence or len(sequence) < sequence_length + 1:
        return samples

    seen_songs = set([e["songId"] for e in sequence])

    negative_pool = [
        song for song in all_songs
        if song["id"] not in seen_songs
    ]

    if len(negative_pool) == 0:
        return samples

    for i in range(sequence_length, len(sequence)):

        context = sequence[i - sequence_length:i]
        target = sequence[i]

        user_tensor = build_user_tensor(context, embedding_length)

        samples.append({
            "user_tensor": user_tensor,
            "song_embedding": target["songEmbedding"],
            "is_positive": 1
        })

        for _ in range(negative_per_positive):

            negative_song = random.choice(negative_pool)

            samples.append({
                "user_tensor": user_tensor,
                "song_embedding": negative_song["embedding"],
                "is_positive": 0
            })

    return samples

File: server/services/recommendations/dataset/buildDataset.js
import { buildUserTensor } from './buildUserTensor.js'

export function buildDataset(
  sequence,
  allSongs,
  sequenceLength = 10,
  negativePerPositive = 1,
  embeddingLength
) {

  const samples = []

  if (!sequence || sequence.length < sequenceLength + 1) {
    return samples
  }

  const seenSongs = new Set(
    sequence.map(e => e.songId)
  )

  const negativePool = allSongs.filter(
    song => !seenSongs.has(song.id)
  )

  if (negativePool.length === 0) return samples

  for (let i = sequenceLength; i < sequence.length; i++) {

    const context = sequence.slice(i - sequenceLength, i)
    const target = sequence[i]

    const userTensor =
      buildUserTensor(context, embeddingLength)

    samples.push({
      userTensor,
      songEmbedding: target.songEmbedding,
      isPositive: 1
    })

    for (let n = 0; n < negativePerPositive; n++) {

      const negativeSong =
        negativePool[
          Math.floor(Math.random() * negativePool.length)
        ]

      samples.push({
        userTensor,
        songEmbedding: negativeSong.embedding,
        isPositive: 0
      })
    }
  }

  return samples
}